common/task/weightedRoundRobinTaskScheduler.go

```diff
@@ -0,0 +1,197 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package task
+
+import (
+	"errors"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/backoff"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+)
+
+type (
+	// WeightedRoundRobinTaskSchedulerOptions configs WRR task scheduler
+	WeightedRoundRobinTaskSchedulerOptions struct {
+		Weights     []int
+		QueueSize   int
+		WorkerCount int
+		RetryPolicy backoff.RetryPolicy
+	}
+
+	weightedRoundRobinTaskSchedulerImpl struct {
+		status       int32
+		numQueues    int
+		taskChs      []chan PriorityTask
+		shutdownCh   chan struct{}
+		notifyCh     chan struct{}
+		dispatcherWG sync.WaitGroup
+		logger       log.Logger
+		metricsScope metrics.Scope
+		options      *WeightedRoundRobinTaskSchedulerOptions
+
+		processor Processor
+	}
+)
+
+const (
+	wRRTaskProcessorQueueSize = 1
+)
+
+var (
+	// ErrTaskSchedulerClosed is the error returned when submitting task to a stopped scheduler
+	ErrTaskSchedulerClosed = errors.New("task scheduler has already shutdown")
+)
+
+// NewWeightedRoundRobinTaskScheduler creates a new WRR task scheduler
+func NewWeightedRoundRobinTaskScheduler(
+	logger log.Logger,
+	metricsScope metrics.Scope,
+	options *WeightedRoundRobinTaskSchedulerOptions,
+) (Scheduler, error) {
+	if len(options.Weights) == 0 {
+		return nil, errors.New("weight is not specified in the scheduler option")
+	}
+
+	numPriorities := len(options.Weights)
+	scheduler := &weightedRoundRobinTaskSchedulerImpl{
+		status:       common.DaemonStatusInitialized,
+		numQueues:    numPriorities,
+		taskChs:      make([]chan PriorityTask, numPriorities),
+		shutdownCh:   make(chan struct{}),
+		notifyCh:     make(chan struct{}, 1),
+		logger:       logger,
+		metricsScope: metricsScope,
+		options:      options,
+		processor: NewParallelTaskProcessor(
+			logger,
+			metricsScope,
+			&ParallelTaskProcessorOptions{
+				QueueSize:   wRRTaskProcessorQueueSize,
```

Comment:
I am confused on QueueSize, should it be options.QueueSize instead of `wRRTaskProcessorQueueSize`?

Reply:
If not, you probably want to renamed it taskChSize or something else

Reply:
Synced offline. Agreed to keep the existing name.